daemon/daemon.go

```diff
@@ -969,3 +977,98 @@ func (d *Daemon) StopIgnoringContainer(id string) {
 	delete(d.ignoredContainers, id)
 	d.ignoredMutex.Unlock()
 }
+
+// listFilterIfs returns a map of interfaces based on the given filter.
+// The filter should take a link and, if found, return the index of that
+// interface, if not found return -1.
+func listFilterIfs(filter func(netlink.Link) int) (map[int]netlink.Link, error) {
+	ifs, err := netlink.LinkList()
+	if err != nil {
+		return nil, err
+	}
+	vethLXCIdxs := map[int]netlink.Link{}
+	for _, intf := range ifs {
+		if idx := filter(intf); idx != -1 {
+			vethLXCIdxs[idx] = intf
+		}
+	}
+	return vethLXCIdxs, nil
+}
+
+// clearCiliumVeths checks all veths created by cilium and removes all that
+// are considered a leftover from failed attempts to connect the container.
+func (d *Daemon) clearCiliumVeths() error {
+	dc := d.dockerClient
+
+	leftVeths, err := listFilterIfs(func(intf netlink.Link) int {
+		// Filter by veth, prefixed with "lxc" and return the index
+		// of the interface.
+		if intf.Type() == "veth" &&
+			strings.HasPrefix(intf.Attrs().Name, "lxc") {
+			return intf.Attrs().Index
+		}
+		return -1
+	})
+
+	if err != nil {
+		return fmt.Errorf("unable to retrieve host network interfaces: %s", err)
+	}
+
+	conts, err := dc.ContainerList(context.Background(), dTypes.ContainerListOptions{})
+	if err != nil {
+		return fmt.Errorf("unable to retrieve all containers running in docker: %s", err)
+	}
+
```

Comment:
I think you can drastically simplify this if we can make the following assumption:
- We never perform networking for an endpoint without the the peer veth being in another network namespace.

If we make that assumption then it becomes extremely simple to figure out whether it is an unused veth pair:
```
for _, l := range leftVeths {
        peerIndex := l.Attrs().ParentIndex
        if _, ok := leftVeths[peerIndex]; !ok && peerIndex {
                // peer of l is not in namespace, schedule l for removal
        }
}
```